### ISSUE TITLE
Initial DAT and Python Implementation

### ISSUE THREAD

**Issue by pkafei** [pkafei/Dat-Python#1]
Dat
![screen shot 2014-10-29 at 8 08 42 am](https://cloud.githubusercontent.com/assets/1463964/4828108/9fc93936-5f7d-11e4-8df4-6c6067bb28ec.png)

Dat datastore in IPython Notebook
![screen shot 2014-10-29 at 8 11 15 am](https://cloud.githubusercontent.com/assets/1463964/4828132/e11952cc-5f7d-11e4-865f-5f10bf8851fe.png)

**Comment by pkafei** [comment]
The example above demonstrates that  DAT as it stands a user can add 'api/csv' to the dat instance on the server 'http://localhost:6461/api/csv' and import it into a Python Pandas Dataframe.

I'm thinking the best way to implement dat functionality into Python is to use the [Pandas IO API](http://pandas.pydata.org/pandas-docs/dev/io.html). Importing in Pandas works like so:

```
example_csv = pd.read_csv()
example_json = pd.read_json()
example_sql = pd.read_sql()
```

I'm imagining dat input into pandas to work in a similar fashion:

```
example_dat = pd.read_dat()
```

The read_dat method will parse the json and dump it into a dataframe. What do you think? Doe this sound like a sound way to implement DAT into Python?

@maxogden @ywyw @willingc @karissa

**Comment by ywyw** [comment]
Interesting, I wonder what the bandwidth and size limits are on this.  Does it work in a streaming sort of fashion or is it memory-bound?  Seems like it should be able to connect to a remote file system, so if the output database is on another server like Digital Ocean, it can export onto something there.

How would you say the sqlalchemy package compares to psycopg2?

**Comment by pkafei** [comment]
@ywyw 
I don't know what the bandwidth and size limits are and I'm assuming Pandas is working in a memory-bound fashion. I've never used a remote file system with Pandas (like a PostGres database hosted on Digital Ocean) but I'm assuming it can be done.

Sqlalchemy is a ORM (allows you to interact with PostGres using Python), and Psycopg2 are drivers that allow you to use raw SQL. Generally in the Python ecosystem there is more support for Sqlalchemy than Pyscopg2.

http://stackoverflow.com/questions/8588126/sqlalchemy-or-psycopg2

**Comment by okdistribute** [comment]
Doing that will require the pandas team to accept our pull request to incorporate dat into their main trunk which could be a tough one.

It might be better to create a dat python module that you can use to import and convert from multiple things, the first being a pandas data frame.

```
import dat
df = dat.read('http://localhost:6767')
```

Later, if pandas team decides to accept, they can create an alias `read_dat` that does the same thing.

Pandas is memory-bound AFAIK.

**Comment by pkafei** [comment]
@karissa 

I also worry that it would be hard to get pandas to accept such a pull request. I think the best first step is to create native Python API in DAT and then perhaps a dat python module.

**Comment by okdistribute** [comment]
yeah i think its a good idea to start with the native apis. a native python api would need to interact with both the on-disk local storage as well as the rest api, but which comes first?

**Comment by okdistribute** [comment]
i'm almost thinking the rest api would come first because they could interact with the local dat using cmd line

**Comment by pkafei** [comment]
@karissa 
I think you're right- a rest api would be a good way to go.

**Comment by okdistribute** [comment]
Cool. You can probably use the read_json under the hood

**Comment by okdistribute** [comment]
(for pandas df stuff)
